microchip: mec: zephyr spi image generation

jvasanth1 opened the pull request:
Added mchp mec zephyr image generator python script. It takes zephyr.bin as input and produces zephyr.mchp.bin.
The default behavior is to not pad to SPI flash size.
(Enable through CONFIG_MCHP_MEC_HEADER=y and CONFIG_MCHP_MEC_HEADER_FLASH_SIZE_256K=y)

zephyr.mchp.bin is composed of:
1. First 4KB contains TAG at offset 0 and header at offset 0x100
2. Offset 0x1000 is the start of zephyr.bin which has been padded to a multiple of 128 bytes.
3. Boot-ROM EC Info Block (128 bytes)
4. Boot-ROM Co-Signature Block (96 bytes)
5. Boot-ROM trailer (160 bytes) contains the SHA-384 digest of 2-4.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

MaureenHelm commented:
Please look at hooking this functionality into `west sign`

jvasanth1 commented:
> Please look at hooking this functionality into `west sign`

This feature is required only for google chrome build.

albertofloyd commented:
@jvasanth1 I assume overall this is orthogonal to using /usr/local/bin/everglades_spi_gen_RomE  whenever west build is run for mec172x board, correct?

jvasanth1 commented:
> @jvasanth1 I assume overall this is orthogonal to using /usr/local/bin/everglades_spi_gen_RomE whenever west build is run for mec172x board, correct?

Yes, it runs separately taking zephyr.bin as input.

keith-zephyr commented on soc/arm/microchip_mec/Kconfig:
```diff
@@ -11,6 +12,210 @@ config SOC_FAMILY
+menuconfig MCHP_MEC_HEADER
+	bool "The output binary with MCHP MEC binary header"
+	help
+	  On Microchip MEC series chip, the ROM code loads firmware image from flash
+	  to RAM using a TAG to locate a Header which specifies the location and
+	  size of the firmware image. Enable this to invoke the mec_spi_gen tool
+	  which generates an SPI image with TAG, Header, and firmware binary. This
+	  tool does not produce a signed image which can be authenticated by the
+	  Boot-ROM. Use the full Microchip SPI image generator program for
+	  authentication and all other Boot-ROM loader features. Refer to the MCHP
+	  EVB boards for an example.
```
Can you elaborate when one should use the `mec_spi_gen` tool vs. the full SPI image generator program `mec172x_spi_gen_lin_x86_64`?

For instance, is there a pin strapping, or a different part number used to run signed and unsigned images? 

Also, if this option can only create unsigned images, it could be clearer to the user if the name was `MCHP_MEC_UNSIGNED_HEADER`

jvasanth1 replied:
`mec_spi_gen` tool should be turned ON only during the chromium build purpose, that's the current usage. 

will rename to `MCHP_MEC_UNSIGNED_HEADER` for clarity.

jvasanth1 replied:
renamed

keith-zephyr commented on soc/arm/microchip_mec/Kconfig:
```diff
@@ -11,6 +12,210 @@ config SOC_FAMILY
 	default "microchip_mec"
 
+menuconfig MCHP_MEC_HEADER
+	bool "The output binary with MCHP MEC binary header"
```
"Create an unsigned output binary with the MCP MEC binary header"

jvasanth1 replied:
ok

keith-zephyr commented on soc/arm/microchip_mec/Kconfig:
```diff
@@ -11,6 +12,210 @@ config SOC_FAMILY
+	range 0 7
+	default 0
+	help
+	  This three bit value corresponds to the QMSPI controllers clock idle and
+	  input/output data phases. Bits[0:2] are CPOL:CPHA_MOSI:CPHA_MISO. Refer
+	  to the data sheet. Default value is 0 corresponding to SPI Mode 0
+	  signalling.
```
Can you document the setting for each of the standard SPI modes (0-3).
For example:
```
Setting this field to 0 selects mode 0, CPOL=0, CPHA_MOSI=0, CPHA_MISO=0
Setting this filed to 7 selects mode 3, CPOL=0, CPHA_MOSI=0, CPHA_MISO=0
```

jvasanth1 replied:
ok

jvasanth1 replied:
updated accordingly

keith-zephyr reviewed (dismissed):
Note, I did not review any of the python source.

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,498 @@
+# Should this be smaller and if so what value?
+# Code could extend into 64KB data
+PLD_MAX_SIZE_MEC172X = 0x68000
+PLD_MAX_SIZE_MEC152X = 0x40000
```
thse are duplicate on what's in CHIP_DICT an unused, can you drop them?

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,498 @@
+    hdr[0] = 0x50
+    hdr[1] = 0x48
+    hdr[2] = 0x43
+    hdr[3] = 0x4d
```
Magic? May be worth its own constant

jvasanth1 replied:
ok

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,498 @@
+    hdr[0xB0:0xE0] = hdr_digest
+
+    return bytes(hdr)
+# end build_header
```
Drop these? Don't think any other Python script in the codebase uses these kind of notations.

jvasanth1 replied:
can you clarify? is the list slice an issue or the return statement?

jvasanth1 replied:
removed function closing comments

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,498 @@
+                        action='store_true',
+                        help="Enable messages to console")
+
+    args = parser.parse_args()
```
move this block to a separate function? (most other scripts do that)

jvasanth1 replied:
ok

jvasanth1 replied:
added separate function

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,498 @@
+
+    args = parser.parse_args()
+
+    VERBOSE_MODE = args.verbose
```
lowercase? It's a normal variable...

jvasanth1 replied:
ok, will update

jvasanth1 replied:
changed to lower case

jvasanth1 commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,498 @@
+HDR_SPI_RD_111 = 0
+HDR_SPI_RD_1118 = 1
+HDR_SPI_RD_1128 = 2
+HDR_SPI_RD_1148 = 3
```
It is user defined constants (0, 1, 2, 3) as per the bootrom documentation.
1st digit - number of I/O pins used to transmit the opcode
2nd digit - number of I/O pins used to transmit the SPI address
3rd digit - number of pins used to read data from flash
4th digit (if present) - dummy clocks between address and data phase

jvasanth1 replied:
ok will add comments to explain

jvasanth1 replied:
added comments to explain

jvasanth1 committed:
microchip: mec: zephyr spi image generation

Added mchp mec zephyr image generator python script. It takes
zephyr.bin as input and produces zephyr.mchp.bin.
The default behavior is to not pad to SPI flash size.
(Enable through CONFIG_MCHP_MEC_UNSIGNED_HEADER=y and
CONFIG_MCHP_MEC_HEADER_FLASH_SIZE_256K=y)

zephyr.mchp.bin is composed of:
1. First 4KB contains TAG at offset 0 and header at offset 0x100
2. Offset 0x1000 is the start of zephyr.bin which has been padded
to a multiple of 128 bytes.
3. Boot-ROM EC Info Block (128 bytes)
4. Boot-ROM Co-Signature Block (96 bytes)
5. Boot-ROM trailer (160 bytes) contains the SHA-384 digest of 2-4.

Signed-off-by: Jay Vasanth <jay.vasanth@microchip.com>

fabiobaltieri commented on soc/arm/microchip_mec/Kconfig:
```diff
@@ -11,6 +12,212 @@ config SOC_FAMILY
+
+config MCHP_MEC_HEADER_SPI_READ_MODE_QUAD
+	bool "SPI flash operates with quad data reading mode"
+endchoice
```
nit: add a space before endchoice (seems like that's how it's commonly done, also on the other instances below)

jvasanth1 replied:
fixed

fabiobaltieri commented on soc/arm/microchip_mec/Kconfig:
```diff
@@ -11,6 +12,212 @@ config SOC_FAMILY
+
+config MCHP_MEC_HEADER_SPI_READ_MODE_QUAD
+	bool "SPI flash operates with quad data reading mode"
+endchoice
```
nit: add a space before endchoice

fabiobaltieri commented on soc/arm/microchip_mec/Kconfig:
```diff
@@ -11,6 +12,212 @@ config SOC_FAMILY
+	bool "SPI flash size 16M Bytes"
+	help
+	  The SPI flash size is 16M Bytes.
+endchoice
```
nit: add a space before endchoice

jvasanth1 replied:
fixed

fabiobaltieri commented on soc/arm/microchip_mec/Kconfig:
```diff
@@ -11,6 +12,212 @@ config SOC_FAMILY
+	  Setting this filed to 7 selects mode 3, CPOL=1, CPHA_MOSI=1, CPHA_MISO=1
+
+config MCHP_HEADER_VERBOSE_OUTPUT
+bool "Build Microchip MEC binary in BIN format"
```
wrong description?

jvasanth1 replied:
fixed

fabiobaltieri commented on soc/arm/microchip_mec/Kconfig:
```diff
@@ -11,6 +12,212 @@ config SOC_FAMILY
+	  Setting this filed to 7 selects mode 3, CPOL=1, CPHA_MOSI=1, CPHA_MISO=1
+
+config MCHP_HEADER_VERBOSE_OUTPUT
+bool "Build Microchip MEC binary in BIN format"
```
wrong description?

jvasanth1 replied:
fixed

fabiobaltieri commented on soc/arm/microchip_mec/Kconfig:
```diff
@@ -11,6 +12,212 @@ config SOC_FAMILY
+
+if MCHP_MEC_UNSIGNED_HEADER
+
+config MCHP_MEC_IMAGE_OUTPUT_BIN
```
Do you need this option? This whole block depends on `MCHP_MEC_UNSIGNED_HEADER`, but then we have this as the condition for calling `mec_spi_gen.py`. That seems redundant and unnecessary since you could go with `MCHP_MEC_IMAGE_OUTPUT_BIN=n` then the options below here are meaningless.

How about keeping only the menuconfig option and use that as condition for calling the builder script?

jvasanth1 replied:
removed `config MCHP_MEC_IMAGE_OUTPUT_BIN`

fabiobaltieri commented on soc/arm/microchip_mec/Kconfig:
```diff
@@ -11,6 +12,212 @@ config SOC_FAMILY
+
+if MCHP_MEC_UNSIGNED_HEADER
+
+config MCHP_MEC_IMAGE_OUTPUT_BIN
```
Do you need this option? This whole block depends on `MCHP_MEC_UNSIGNED_HEADER`, but then we have this as the condition for calling `mec_spi_gen.py`. That seems redundant and unnecessary since you could go with `MCHP_MEC_IMAGE_OUTPUT_BIN=n` then the options below here are meaningless.

How about keeping only the menuconfig option and use that as condition for calling the builder script?

jvasanth1 replied:
removed `config MCHP_MEC_IMAGE_OUTPUT_BIN`

fabiobaltieri commented on soc/arm/microchip_mec/Kconfig:
```diff
@@ -11,6 +12,212 @@ config SOC_FAMILY
+	default "fast" if MCHP_MEC_HEADER_SPI_SLEW_RATE_FAST
+
+config MCHP_MEC_HEADER_FLASH_SPI_MODE
+	int
```
Does this need a description? It can't be set to anything other than the default otherwise.

fabiobaltieri replied:
Since the option has no description it cannot be set to anything other than the default (unless the Kconfig entry is overridden somewhere else). Trying to set it gives you a build error:

```
error: MCHP_MEC_HEADER_FLASH_SPI_MODE (defined at soc/arm/microchip_mec/Kconfig:201) is assigned in
a configuration file, but is not directly user-configurable (has no prompt). It gets its value
indirectly from other symbols. See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_MCHP_MEC_HEADER_FLASH_SPI_MODE and/or look
up MCHP_MEC_HEADER_FLASH_SPI_MODE in the menuconfig/guiconfig interface. The Application Development
Primer, Setting Configuration Values, and Kconfig - Tips and Best Practices sections of the manual
might be helpful too.
```

See https://docs.zephyrproject.org/2.6.0/guides/kconfig/setting.html#visible-and-invisible-kconfig-symbols

jvasanth1 replied:
doesn't it have the description under help?

```
config MCHP_MEC_HEADER_FLASH_SPI_MODE
	int
	range 0 7
	default 0
	help
	  This three bit value corresponds to the QMSPI controllers clock idle and
	  input/output data phases. Bits[0:2] are CPOL:CPHA_MOSI:CPHA_MISO. Refer
	  to the data sheet. Default value is 0 corresponding to SPI Mode 0
	  signalling.
	  Setting this field to 0 selects mode 0, CPOL=0, CPHA_MOSI=0, CPHA_MISO=0
	  Setting this filed to 7 selects mode 3, CPOL=1, CPHA_MOSI=1, CPHA_MISO=1
```

fabiobaltieri replied:
Yeah it has a help text but what matters it does not have a prompt (the short description after the type, it's just `int` rather than, for example, `int "Header Flash SPI mode"`). Try to build a project overriding that config option, in my case I'm building one example and added `CONFIG_MCHP_MEC_HEADER_FLASH_SPI_MODE=3` in the prj.conf file, it should fail the build.

jvasanth1 replied:
ok, I got your point. I will add the prompt.

fabiobaltieri commented on soc/arm/microchip_mec/Kconfig:
```diff
@@ -11,6 +12,212 @@ config SOC_FAMILY
+	default "fast" if MCHP_MEC_HEADER_SPI_SLEW_RATE_FAST
+
+config MCHP_MEC_HEADER_FLASH_SPI_MODE
+	int
```
Does this need a description? It can't be set to anything otherwise.

jvasanth1 replied:
added prompt

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,510 @@
+    if verbose_mode:
+        print("MEC SPI Gen done")
+
+# end main
```
drop this line

jvasanth1 replied:
removed

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,510 @@
+    if verbose_mode:
+        print("MEC SPI Gen done")
+
+# end main
```
drop this line

jvasanth1 replied:
removed

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,510 @@
+            print("ERROR: Specify output binary file name with -o")
+        sys.exit(-1)
+
+    chip = CHIP_DEFAULT
```
This should not be needed, the flag already has a default.

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,510 @@
+            print("ERROR: Specify output binary file name with -o")
+        sys.exit(-1)
+
+    chip = CHIP_DEFAULT
```
This should not be needed, the flag already has a default.

jvasanth1 replied:
the idea of assigning chip is that if the args passed isn't 'mec172x' or 'mec152x', line 333 will not be executed and we could continue with the default value after reporting warning.

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,510 @@
+        chip = args.chip
+    else:
+        if verbose_mode:
+            print("ERROR: Unknown chip {0}. Use default chip: {1}".format(args.chip, CHIP_DEFAULT))
```
This pattern (discard an explicit flag, fall back to default, only warn on verbose=True) seems like asking for trouble. If the flag value is not valid make the script fail explicitly and return the error message.

Same for the other options below, verbosity check should not be part of flag argument validation.

Also, can you use the `choices` option for validation rather than a custom check here? https://docs.python.org/3/library/argparse.html#choices

jvasanth1 replied:
Added `choices` option for validation

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,510 @@
+        chip = args.chip
+    else:
+        if verbose_mode:
+            print("ERROR: Unknown chip {0}. Use default chip: {1}".format(args.chip, CHIP_DEFAULT))
```
This pattern (discard an explicit flag, fall back to default, only warn on verbose=True) seems like asking for trouble. If the flag value is not valid make the script fail explicitly and return an error.

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,510 @@
+    return bytes(hdr)
+
+def parse_args():
+
```
nit: drop the empty line at the beginning of the function

jvasanth1 replied:
removed

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,510 @@
+
+def main():
+    """MEC SPI Gen"""
+
```
nit: drop the empty line

jvasanth1 commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,510 @@
+    return bytes(hdr)
+
+def parse_args():
+
```
removed

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,476 @@
+#           ARM vector table at beginning of payload and reset handler
+#           address is at offset 4 of payload.
+#         Payload length, must be multiple of PLD_GRANULARITY
+#
```
How would you feel about moving these comments in the function docstring? That's how it's done traditionally in Python, Doxygen would potentially pick them up as well which is kinda nice (see the "Comment blocks in Python" in https://www.doxygen.nl/manual/docblocks.html).

Applies to this and the two functions above.

jvasanth1 replied:
moved comments to function docstring.

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,476 @@
+        sys.exit(-1)
+
+    if args.outfilename is None:
+        if verbose_mode:
```
same as above

jvasanth1 replied:
dropped verbose mode for error prints

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,476 @@
+
+    # Do we have enough space for 4KB block containing TAG and Header, padded FW binary,
+    # EC Info Block, Co-Sig Block, and Trailer?
+    if indata_len > (spi_size - (PLD_SPI_LOC + EC_INFO_BLOCK_SIZE + COSIG_SIZE + TRAILER_SIZE)):
```
`spi_size - (PLD_SPI_LOC + EC_INFO_BLOCK_SIZE + COSIG_SIZE + TRAILER_SIZE` could probably use an intermediate varaible for clarity, something like available_size? So that it can be understood without the comment

jvasanth1 replied:
added intermediate variable

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,476 @@
+    # Do we have enough space for 4KB block containing TAG and Header, padded FW binary,
+    # EC Info Block, Co-Sig Block, and Trailer?
+    if indata_len > (spi_size - (PLD_SPI_LOC + EC_INFO_BLOCK_SIZE + COSIG_SIZE + TRAILER_SIZE)):
+        if verbose_mode:
```
drop the condition, always print the error before exiting

jvasanth1 replied:
dropped verbose mode for error prints

jvasanth1 commented on soc/arm/microchip_mec/Kconfig:
```diff
@@ -11,6 +12,210 @@ config SOC_FAMILY
+	  Setting this field to 0 selects mode 0, CPOL=0, CPHA_MOSI=0, CPHA_MISO=0
+	  Setting this filed to 7 selects mode 3, CPOL=1, CPHA_MOSI=1, CPHA_MISO=1
+
+config MCHP_HEADER_VERBOSE_OUTPUT
```
fixed

jvasanth1 commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,476 @@
+        print("  Verbose = {0}".format(args.verbose))
+
+    if args.infilename is None:
+        if verbose_mode:
```
dropped verbose mode for error prints

jvasanth1 commented:
@fabiobaltieri fixed the review comments, could you help to approve?

jvasanth1 commented:
@fabiobaltieri could you approve this PR?

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,476 @@
+# -*- coding: utf-8 -*-
```
Add add python shebang? (see the other python scripts in the repository)

And I think you don't need the utf-8 tag with python3 so that should be removed.

jvasanth1 replied:
updated

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,476 @@
+                        type=str,
+                        dest="infilename",
+                        default="zephyr.bin",
+                        help="Input firmware binary file path/name. Default=zephyr.bin")
```
The library supports having the default in the helptext without repeating it, have a look at https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/gen_cfb_font_header.py#L218 for an example, can you use the same helptext style?

jvasanth1 replied:
updated

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,476 @@
+def build_header(chip, spi, hdr_spi_loc, pld_spi_loc, pld_entry_addr, pld_len):
+    """Build MEC152x/MEC172x Boot-ROM SPI image header
+
+    Inputs: Header location in SPI Image
+            Payload(FW binary) location in SPI Image
+            Payload load address in MEC172x SPI SRAM
+            Payload entry point address: 0 instructs Boot-ROM to assume
+            ARM vector table at beginning of payload and reset handler
+            address is at offset 4 of payload.
+            Payload length, must be multiple of PLD_GRANULARITY
```
This is hard to follow in the current form, does not seem to match the order of the arguments, could you make it clear what the comment refer to? I would suggest the Google standard (see https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/utils/pinctrl_nrf_migrate.py#L306 for an example) but really anthing that removes ambiguity and is used consistently throughout the script would do.

jvasanth1 replied:
updated comments

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,476 @@
+    if verbose_mode:
+        print("Padded FW binary: length = {0}".format(len(indata)))
+
+    indata_len = len(indata)
```
This is a bit confusing, how about: move it in the padding block as "indata_len += pad_len", reuse it as is in the print above (no need to redo len(indata))

jvasanth1 replied:
updated

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,476 @@
+    # EC Info Block, Co-Sig Block, and Trailer?
+    mec_add_info_size = PLD_SPI_LOC + EC_INFO_BLOCK_SIZE + COSIG_SIZE + TRAILER_SIZE
+    if indata_len > (spi_size - mec_add_info_size):
+        print("ERROR: FW binary exceeds flash size!")
```
May be worth printing something more with the error, maybe indata_len and spi_size. Imagine this failing, one would could use verbose to get more info but then you'd have to figure out what command line the build system used. Having the relevant data with the error would  help out.

jvasanth1 replied:
updated

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,476 @@
+        print_bytes("TAG", tag)
+        print("Build Header at {0}: Load Address = 0x{1:0x} Entry Point Address = 0x{2:0x}".format(HDR_SPI_LOC, PLD_LOAD_ADDR, entry_point))
+
+    spi_dict = { "spi_freq_mhz": spi_freq_mhz, "spi_mode": spi_mode,
```
Should this be something like `spi_config`? The variable name should describe the intent not the data type. Same in build_header() where it just says `spi`.

jvasanth1 replied:
updated accordingly

fabiobaltieri commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,476 @@
+        print("SHA-384 digest (paddedFW || ec_info_block || cosig)")
+        print_bytes("digest", image_digest)
+
+    spi_bufs = []
```
Would it make more sense to use a dictionary here? Then you could iterate it over it with just sorted(spi_bufs) and unpack the values directly using items(), which I think would make the code below a bit easier to understand.

jvasanth1 replied:
prefer to use list.

fabiobaltieri reviewed (commented):
Just few leftover cosmetic comments on my side.

jvasanth1 commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,481 @@
+
+    print("}")
+
+def Crc8(crc, data):
```
updated

jvasanth1 commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,481 @@
+
+    if verbose_mode:
+        print_bytes("TAG", tag)
+        print("Build Header at {0}: Load Address = 0x{1:0x} Entry Point Address = 0x{2:0x}".format(HDR_SPI_LOC, PLD_LOAD_ADDR, entry_point))
```
updated

jvasanth1 commented on soc/arm/microchip_mec/common/spigen/mec_spi_gen.py:
```diff
@@ -0,0 +1,481 @@
+        print_bytes("TAG", tag)
+        print("Build Header at {0}: Load Address = 0x{1:0x} Entry Point Address = 0x{2:0x}".format(HDR_SPI_LOC, PLD_LOAD_ADDR, entry_point))
+
+    spi_config_info = { "spi_freq_mhz": spi_freq_mhz, "spi_mode": spi_mode,
```
updated

fabiobaltieri commented:
> @fabiobaltieri can you help to approve? Thanks

Sorry I'm out on holiday, @jackrosenthal would you have time to take a look? I'll be back next week otherwise.

jvasanth1 commented:
@keith-zephyr can you help to approve?
